### ISSUE TITLE
Basic blocks not solved correctly

### ISSUE THREAD

**Issue by roj4s** [radare/radare2#7170]
A basic block is a straight-line sequence of code with only one entry point and only one exit.
Seems like Radare2 don't construct basic blocks correctly, it don't consider call opcodes. See below commands outputs. Maybe I'm missing something here. Thanks in advance.

for i in p.cmdj('afbj @main'):
    print i
    
{u'inputs': 0, u'addr': 4205056, u'ninstr': 35, u'outputs': 1, u'traced': False, u'fail': 4205243, u'size': 187}
{u'inputs': 1, u'addr': 4205243, u'ninstr': 2, u'outputs': 2, u'traced': False, u'jump': 4205295, u'fail': 4205248, u'size': 5}
{u'inputs': 1, u'addr': 4205248, u'ninstr': 2, u'outputs': 2, u'traced': False, u'jump': 4205258, u'fail': 4205253, u'size': 5}
{u'inputs': 1, u'addr': 4205253, u'ninstr': 1, u'outputs': 0, u'traced': False, u'size': 5}
{u'inputs': 1, u'addr': 4205258, u'ninstr': 4, u'outputs': 1, u'traced': False, u'fail': 4205276, u'size': 18}
{u'inputs': 1, u'addr': 4205276, u'ninstr': 3, u'outputs': 0, u'traced': False, u'size': 19}
{u'inputs': 1, u'addr': 4205295, u'ninstr': 4, u'outputs': 0, u'traced': False, u'size': 22}
for i in p.cmdj('pdrj @main'):
    print i['offset']," ",i['opcode']
    
4205056   push r15
4205058   push r14
4205060   push r13
4205062   push r12
4205064   push rbp
4205065   push rbx
4205066   mov ebx, edi
4205068   mov rbp, rsi
4205071   sub rsp, 0x388
4205078   mov rdi, qword [rsi]
4205081   mov rax, qword fs:[0x28]
4205090   mov qword [local_378h], rax
4205098   xor eax, eax
4205100   call 0x40db00
4205105   mov esi, 0x419b21
4205110   mov edi, 6
4205115   call 0x402840
4205120   mov esi, 0x416611
4205125   mov edi, 0x4165fa
4205130   call 0x4024b0
4205135   mov edi, 0x4165fa
4205140   call 0x402470
4205145   mov edi, 0x40a320
4205150   mov dword [0x0061e580], 2
4205160   call 0x413c60
4205165   movabs rax, 0x8000000000000000
4205175   mov dword [0x0061f030], 0
4205185   mov byte [0x0061f0d8], 1
4205192   mov qword [0x0061f190], rax
4205199   mov eax, dword [0x0061e56c]
4205205   mov qword [0x0061f1a0], 0
4205216   mov qword [0x0061f197], -1
4205227   mov byte [0x0061f0f8], 0
4205234   cmp eax, 2
4205237   je 0x403328
4205243   cmp eax, 3
4205246   je 0x402aef
4205248   sub eax, 1
4205251   je 0x402aca
4205253   call 0x402370
4205258   mov edi, 1
4205263   call 0x4023e0
4205268   test eax, eax
4205270   je 0x4035a6
4205276   mov dword [0x0061f158], 2
4205286   mov byte [0x0061f0f8], 1
4205293   jmp 0x402b05
4205295   mov esi, 7
4205300   xor edi, edi
4205302   mov dword [0x0061f158], 0
4205312   call 0x40eca0

**Comment by Maijin** [comment]
2 - Use markdown for code

**Comment by radare** [comment]
> On 28 Mar 2017, at 14:16, Luis Miguel Rojas Aguilera <notifications@github.com> wrote:
> 
> A basic block is a straight-line sequence of code with only one entry point and only one exit.
> 

False
> Seems like Radare2 don't construct basic blocks correctly, it don't consider call opcodes. See below commands outputs. Maybe I'm missing something here. Thanks in advance.
> 

Yes it does. Call instructions dont break basic blocks.

**Comment by roj4s** [comment]
So call instructions don't change execution flow? 
There are also jumps and je not being considered ...
 Ok I'm not expert in this field ... I'm trying to construct a graph where nodes are instructions and edges connects nodes that act with the same data references (regs, memory ...) May I just follow the output of pdr command ... Thanks in advance ...

**Comment by radare** [comment]
In r1 theres an option to do that but in the re field, calls are not considered a basic block break.

In fact there are callgraphs which do that and they use to be already quite big. So making those at basic block level will make that even more difficult to read

**Comment by radare** [comment]
question: do we really want the anal.callsplit back from r1?

**Comment by roj4s** [comment]
Maybe you can give me an advice. 
  I'm using the Worklist approach for data flow analysis for what a control-flow graph is necessary so the output of afb would do it.  My strategy is to iterate throug the outputs of pdr to get instructions' memory space writen and read drawing edges between instructions that uses the same memory references (regs etc...).  And then do that for each function.

**Comment by XVilka** [comment]
@radare callsplit option could be handy. There is even a function for that in LLVM,

**Comment by alvarofe** [comment]
This is gonna produce long cfg per function and mess everything up imho. In which case would be handy?

**Comment by Pamplemousse** [comment]
TL;DR: Not breaking BB at callsites prevent some static analyses to be implemented in `radare2`.

---

@radare [your comment](#issuecomment-289875878-permalink) is very misleading...

>  A basic block is a straight-line sequence of code with only one entry point and only one exit.

Actually, this definition [is correct](https://en.wikipedia.org/wiki/Basic_block).
And it is the one textbook analyses will rely on: def-use, reaching definitions...

> In r1 theres an option to do that but in the re field, calls are not considered a basic block break.

Well, it depends what you are trying to do:
  * True, it is clearer at display (see IDA or Ghidra) not to break on calls;
  * But as mentioned, for certain analyses, you **need** to consider that `call` break basic blocks, because **they change the control flow** of a program (this is what angr does, [for this exact reason](https://docs.angr.io/introductory-errata/faq#why-is-angrs-cfg-different-from-idas)).

**Comment by karliss** [comment]
I agree with with @Pamplemousse  it all depends. From practical perspective is anyone currently planning to implement an analysis and is being blocked by r2 not splitting BB at callsites? Would doing that allow making better r2 or Cutter angr plugin?

Would it be a significant problem to do the splitting during data preparation for the specific analysis, compared to the complexity of implementing the analysis itself?

If you want to split BB at callsites you probably also want the edges. Generating the edges for virtual function calls, return instruction -> BB after call instruction, considering the case where return instruction on stack is overwritten are non-trivial and depends on the needs of analysis. I would say it's better left as a task of code doing cross function analysis.

Taking it to the extreme even a mov(ldr, str) instruction can be considered potential change in control flow due to it generating an interrupt. Any instruction generating an interrupt can. For the purpose of analyzing embedded firmware such control flow wouldn't be exceptional, and normal code flow might rely on certain changes in state to be done by interrupt handlers. If you wanted to properly analyze it for proving correctness or searching a race condition you would have take such jumps into account. Even worse a timer or external input interrupt could cause a control flow change at almost any instruction. This isn't exclusive to embedded systems, obfuscated malware can also rely on additional control flow changes through operating system by triggering interrupts to hinder static and dynamic analysis. All of that potentially prevents (not really) some static analysis to be implemented. I hope everyone agrees it's not really practical for r2 splitting BB after every mov instruction. Where is the line?

On the other end of extreme - call instruction can be considered a big fat instruction that pops arguments and pushes result, especially for platforms like JVM where call instruction effect on stack  is very limited and instructions can't access data across the stack frames.

**Comment by radare** [comment]
I dont think my opinion counts at all and also i think this has been solved already a while ago with florians blocks and defragger’s aab

**Comment by Pamplemousse** [comment]
I am very new to `radare2`, so I don't see what "florians blocks" or "defragger's aab" refer too...

Not sure this is what you refer to:
```
[0x00401060]> aa?
[...]
| aab                 abb across bin.sections.rx

[0x00401060]> ab?
[...]
| abb [length]         analyze N bytes and extract basic blocks
```

Could you provide links to the relevant part of the documentation please?
Or guidance how to use say to recover BB from a given address?

Thanks for your help :)
